src/runtime-rs/crates/hypervisor/src/qemu/inner.rs

```diff
@@ -0,0 +1,133 @@
+use anyhow::Result;
+
+use crate::{HypervisorConfig, VcpuThreadIds};
+use kata_types::capabilities::{Capabilities, CapabilityBits};
+
+const VSOCK_SCHEME: &str = "vsock";
+const VSOCK_AGENT_CID: u32 = 3;
+const VSOCK_AGENT_PORT: u32 = 1024;
+
+unsafe impl Send for QemuInner {}
+unsafe impl Sync for QemuInner {}
+
+pub struct QemuInner {
+    config: HypervisorConfig,
+}
+
+impl QemuInner {
+    pub fn new() -> QemuInner {
+        QemuInner {
+            config: Default::default(),
+        }
+    }
+
+    pub(crate) async fn prepare_vm(&mut self, _id: &str, _netns: Option<String>) -> Result<()> {
+        info!(sl!(), "Preparing QEMU VM");
+        Ok(())
+    }
+
+    pub(crate) async fn start_vm(&mut self, _timeout: i32) -> Result<()> {
+        info!(sl!(), "Starting QEMU VM");
+
+        let mut command = std::process::Command::new(&self.config.path);
+
+        command
+            .arg("-kernel")
+            .arg(&self.config.boot_info.kernel)
+            .arg("-m")
+            .arg(format!("{}M", &self.config.memory_info.default_memory))
+            .arg("-initrd")
+            .arg(&self.config.boot_info.initrd)
+            .arg("-vga")
+            .arg("none")
+            .arg("-nodefaults")
+            .arg("-nographic");
+
+        command.spawn()?;
+
+        Ok(())
+    }
+
+    pub(crate) fn stop_vm(&mut self) -> Result<()> {
+        info!(sl!(), "Stopping QEMU VM");
+        todo!()
+    }
+
+    pub(crate) fn pause_vm(&self) -> Result<()> {
+        info!(sl!(), "Pausing QEMU VM");
+        todo!()
+    }
+
+    pub(crate) fn resume_vm(&self) -> Result<()> {
+        info!(sl!(), "Resuming QEMU VM");
+        todo!()
+    }
+
+    pub(crate) async fn save_vm(&self) -> Result<()> {
+        todo!()
+    }
+
+    pub(crate) async fn get_agent_socket(&self) -> Result<String> {
+        info!(sl!(), "QemuInner::get_agent_socket()");
+        const VSOCK_SCHEME: &str = "vsock";
```

Comment:
maybe #todo is needed here for future vsock related support work?

Reply:
I'm actually working on this already as we speak ;-) I'm trying to employ the vsock code added by PR  #5691 to support dynamic allocation of CID, however qemu somehow disagrees with the result so far. Trying to figure it out.

Reply:
So maybe this work could be done in later PR and we could first add a `#todo` here?

Reply:
I'm not sure if you mean a todo as in a comment or a todo as in the macro.  I can sure add a comment but I think calling `todo!()` here would not be helpful as it would make it impossible to run and test the code. As is, this code is already able to launch a VM although with a limitation that there can be just a single such VM on a host - which is fine for this stage of development. Once the dynamic CID allocation work succeeds the limitation will be lifted.

Reply:
I mean the todo in the comment so that next time when we read through the code we could know that this part could be optimized and what to optimize :)

Example for the todo comments: https://github.com/kata-containers/kata-containers/blob/main/src/agent/src/netlink.rs#L519

Reply:
Okay, I'll add the comment.